File: TsinghuaNet.Avalonia/NetCommand.cs
using System;
using System.ComponentModel;
using System.Threading.Tasks;
using System.Windows.Input;

namespace TsinghuaNet.Avalonia
{
    abstract class NetCommand : ICommand
    {
        protected MainViewModel model;
        protected bool executing;
        public NetCommand(MainViewModel model)
        {
            this.model = model;
            model.PropertyChanged += OnModelPropertyChanged;
        }

        private void OnModelPropertyChanged(object sender, PropertyChangedEventArgs e)
        {
            if (e.PropertyName == "State")
            {
                OnCanExecuteChanged(EventArgs.Empty);
            }
        }

        public event EventHandler CanExecuteChanged;
        protected virtual void OnCanExecuteChanged(EventArgs e) => CanExecuteChanged?.Invoke(this, e);

        public bool CanExecute(object parameter)
        {
            return !executing && model.State > NetState.Unknown && model.State <= NetState.Auth6;
        }

        public async void Execute(object parameter)
        {
            executing = true;
            OnCanExecuteChanged(EventArgs.Empty);
            try
            {
                var helper = model.GetHelper();
                await ExecuteAsync(helper);
            }
            finally
            {
                executing = false;
                OnCanExecuteChanged(EventArgs.Empty);
            }
        }

        protected abstract Task ExecuteAsync(IConnect helper);
    }

    class LoginCommand : NetCommand
    {
        public LoginCommand(MainViewModel model) : base(model) { }

        protected override async Task ExecuteAsync(IConnect helper)
        {
            if (helper != null)
                await helper.LoginAsync();
            await model.RefreshAsync(helper);
        }
    }

    class LogoutCommand : NetCommand
    {
        public LogoutCommand(MainViewModel model) : base(model) { }

        protected override async Task ExecuteAsync(IConnect helper)
        {
            if (helper != null)
                await helper.LogoutAsync();
            await model.RefreshAsync(helper);
        }
    }

    class RefreshCommand : NetCommand
    {
        public RefreshCommand(MainViewModel model) : base(model) { }

        protected override async Task ExecuteAsync(IConnect helper)
        {
            await model.RefreshAsync(helper);
        }
    }
}

File: TsinghuaNet.Avalonia/NetCommand.vb
Imports System.ComponentModel
Imports System.Windows.Input
Imports TsinghuaNet.Helper

Public MustInherit Class NetCommand
    Implements ICommand

    Protected ReadOnly model As MainViewModel
    Protected executing As Boolean
    Public Sub New(model As MainViewModel)
        Me.model = model
        AddHandler model.PropertyChanged, AddressOf OnModelPropertyChanged
    End Sub

    Private Sub OnModelPropertyChanged(sender As Object, e As PropertyChangedEventArgs)
        If e.PropertyName = "State" Then
            OnCanExecuteChanged(EventArgs.Empty)
        End If
    End Sub

    Public Async Sub Execute(parameter As Object) Implements ICommand.Execute
        executing = True
        OnCanExecuteChanged(EventArgs.Empty)
        Try
            Dim helper = model.GetHelper()
            Await ExecuteAsync(helper)
        Finally
            executing = False
            OnCanExecuteChanged(EventArgs.Empty)
        End Try
    End Sub

    Protected MustOverride Function ExecuteAsync(helper As IConnect) As Task

    Public Event CanExecuteChanged As EventHandler Implements ICommand.CanExecuteChanged
    Protected Overridable Sub OnCanExecuteChanged(e As EventArgs)
        RaiseEvent CanExecuteChanged(Me, e)
    End Sub

    Public Function CanExecute(parameter As Object) As Boolean Implements ICommand.CanExecute
        Return Not executing AndAlso model.State > NetState.Unknown AndAlso model.State <= NetState.Auth6
    End Function
End Class

Class LoginCommand
    Inherits NetCommand

    Public Sub New(model As MainViewModel)
        MyBase.New(model)
    End Sub

    Protected Overrides Async Function ExecuteAsync(helper As IConnect) As Task
        If helper IsNot Nothing Then
            Await helper.LoginAsync()
        End If
        Await model.RefreshAsync(helper)
    End Function
End Class

Class LogoutCommand
    Inherits NetCommand

    Public Sub New(model As MainViewModel)
        MyBase.New(model)
    End Sub

    Protected Overrides Async Function ExecuteAsync(helper As IConnect) As Task
        If helper IsNot Nothing Then
            Await helper.LogoutAsync()
        End If
        Await model.RefreshAsync(helper)
    End Function
End Class

Class RefreshCommand
    Inherits NetCommand

    Public Sub New(model As MainViewModel)
        MyBase.New(model)
    End Sub

    Protected Overrides Async Function ExecuteAsync(helper As IConnect) As Task
        Await model.RefreshAsync(helper)
    End Function
End Class
